src/test/java/net/unicon/lti/utils/lti/LTI3RequestTest.java

```diff
@@ -183,26 +193,138 @@ public void testLTI3RequestWithInvalidNonce() {
+        IllegalStateException exception = Assertions.assertThrows(IllegalStateException.class, () -> {new LTI3Request(req, ltiDataService, false, "1234", jwsClaims);});
 
         verify(platformDeploymentRepository).findByIssAndClientIdAndDeploymentId(any(String.class), any(String.class), any(String.class));
         verify(claims).get(eq(LtiStrings.LTI_VERSION), eq(String.class));
         verify(claims).get(eq(LtiStrings.LTI_MESSAGE_TYPE), eq(String.class));
         verify(claims).get(eq(LtiStrings.LTI_NONCE), eq(String.class));
+        assertEquals("Nonce error: Unknown or already used nonce.", exception.getMessage());
+
+    }
+
+    @Test
+    public void testLTI3RequestWithInvalidNonceCookieless() {
+        Map<String,Object> testMap = new HashMap<>();
+        testMap.put("id","value1");
+        List<String> testList = new ArrayList<>();
+        testList.add("key1");
+        when(req.getParameter("id_token")).thenReturn(ID_TOKEN);
+        when(req.getParameter("cookies")).thenReturn("false");
+        when(platformDeploymentRepository.findByIssAndClientIdAndDeploymentId(any(String.class), any(String.class), any(String.class))).thenReturn(platformDeploymentList);
+        when(claims.getIssuer()).thenReturn(SAMPLE_ISS);
+        when(claims.getAudience()).thenReturn(SAMPLE_CLIENT_ID);
+        when(claims.getExpiration()).thenReturn(new Date());
+        when(claims.getIssuedAt()).thenReturn(new Date());
+        when(claims.getSubject()).thenReturn(SAMPLE_USER);
+        when(claims.containsKey(eq(LtiStrings.LTI_LINK))).thenReturn(true);
+        when(claims.get(eq(LtiStrings.LTI_LINK), eq(Map.class))).thenReturn(testMap);
+        when(claims.containsKey(eq(LtiStrings.LTI_ROLES))).thenReturn(true);
+        when(claims.get(eq(LtiStrings.LTI_ROLES), eq(List.class))).thenReturn(testList);
+        when(claims.containsKey(eq(LtiStrings.LTI_DEPLOYMENT_ID))).thenReturn(true);
+        when(claims.get(eq(LtiStrings.LTI_DEPLOYMENT_ID))).thenReturn(SAMPLE_DEPLOYMENT_ID);
+        when(claims.get(eq(LtiStrings.LTI_DEPLOYMENT_ID), eq(String.class))).thenReturn(SAMPLE_DEPLOYMENT_ID);
+        when(claims.containsKey(eq(LtiStrings.LTI_VERSION))).thenReturn(true);
+        when(claims.get(eq(LtiStrings.LTI_VERSION), eq(String.class))).thenReturn(LtiStrings.LTI_VERSION_3);
+        when(claims.containsKey(eq(LtiStrings.LTI_MESSAGE_TYPE))).thenReturn(true);
+        when(claims.get(eq(LtiStrings.LTI_MESSAGE_TYPE), eq(String.class))).thenReturn(LtiStrings.LTI_MESSAGE_TYPE_RESOURCE_LINK);
+        when(claims.get(eq(LtiStrings.LTI_MESSAGE_TYPE))).thenReturn(LtiStrings.LTI_MESSAGE_TYPE_RESOURCE_LINK);
+        when(claims.get(eq(LtiStrings.LTI_VERSION), eq(String.class))).thenReturn(LtiStrings.LTI_VERSION_3);
+        mockHttpSession.setAttribute("lti_nonce", Arrays.asList("5d04cb12f45df6ee373c42a1ca4cdbe08e2bfa8e5f7c662aca3f6560687fdc47"));
```

Comment:
Does this still need to get set with the cookieless version?

Reply:
Yes, if it is cookieless it just won't be checked. We always set the cookies just in case.